packages/next/src/build/webpack/plugins/slow-module-detection-plugin.ts

```diff
@@ -0,0 +1,240 @@
+  const debugId = module.debugId
+  return String(debugId)
+}
+
+const getModuleDisplayName = (module: Module): string | undefined => {
+  const resourcePath =
+    'resource' in module && typeof module.resource === 'string'
+      ? module.resource
+      : undefined
+
+  if (!resourcePath) {
+    return undefined
+  }
+
+  let displayPath = resourcePath.replace(process.cwd(), '.')
+
+  const nodeModulesMatch = displayPath.match(NODE_MODULES_PATH_PATTERN)
+  if (nodeModulesMatch) {
+    return nodeModulesMatch[1]
+  }
+
+  return displayPath
+}
+
+// Truncates long paths by preserving the beginning and end portions while adding ellipsis in the middle
+// Example (with PATH_TRUNCATION_LENGTH = 100):
+//   Input:  "/very/long/path/node_modules/.pnpm/some-package@1.0.0/node_modules/deep/nested/file/that/goes/on/for/quite/some/time.js"
+//   Output: "/very/long/path/node_modules/.pnpm/...ested/file/that/goes/on/for/quite/some/time.js"
+function truncatePath(path: string): string {
+  if (path.length <= PATH_TRUNCATION_LENGTH) return path
+
+  // Take the first third of the max length from the start
+  const startSegment = path.slice(0, PATH_TRUNCATION_LENGTH / 3)
+
+  // Take the last two-thirds of the max length from the end
+  const endSegment = path.slice((-PATH_TRUNCATION_LENGTH * 2) / 3)
+
+  return `${startSegment}...${endSegment}`
+}
+
+class ModuleBuildTimeAnalyzer {
+  private pendingModules: Module[] = []
+  private modules = new Map<string, Module>()
+  private moduleParents = new Map<Module, Module>()
+  private moduleChildren = new Map<Module, Map<string, Module>>()
+  private isFinalized = false
+  private slowModuleThresholdMs: number
+  private moduleBuildTimes = new WeakMap<Module, number>()
+
+  constructor(private options: ModuleBuildTimeAnalyzerOptions) {
+    this.slowModuleThresholdMs = options.slowModuleThresholdMs
+  }
+
+  recordModuleBuildTime(module: Module, duration: number) {
+    if (this.isFinalized) {
+      throw new Error(
+        `Invariant (SlowModuleDetectionPlugin): Module is recorded after the report is generated. This is a Next.js internal bug.`
+      )
+    }
+
+    if (duration < this.slowModuleThresholdMs) {
+      return // Skip fast modules
+    }
+
+    this.moduleBuildTimes.set(module, duration)
+    this.pendingModules.push(module)
+  }
+
+  /**
+   * For each slow module, traverses up the dependency chain to find all ancestor modules.
+   * Builds a directed graph where:
+   * 1. Each slow module and its ancestors become nodes
+   * 2. Edges represent "imported by" relationships
+   * 3. Root nodes are entry points with no parents
+   *
+   * The resulting graph allows us to visualize the import chains that led to slow builds.
+   */
+  private prepareReport(compilation: Compilation) {
+    for (const module of this.pendingModules) {
+      // Track visited modules to detect circular dependencies
+      const chain: Module[] = []
+      const visitedModules = new Set<Module>()
```

Comment:
```suggestion
      const visitedModules = new Set<Module>()
```

You don't need `chain`, since `visitedModules` already stores the same list of modules. `Set`s are in insertion order in JS.